Add Afterpay (Clearpay) to the available payment methods

wjrosa opened the pull request:
<!--
Did I add a title? A descriptive, yet concise, title.
-->

<!--
Issue: Link to the GitHub issue this PR addresses (if appropriate).
-->

Fixes #3045

## Changes proposed in this Pull Request:

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

This PR introduces the support for the payment method Afterpay (or Clearpay in the UK) with the "New checkout experience" feature enabled as part of the BNPL methods project. 

Afterpay offers a worldwide payment solution enabling customers to divide purchases into four interest-free payments or opt for longer-term, interest-bearing monthly installments (available in the US only).

<!--
Questions for the PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

## Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

- Checkout to this branch on your local environment (`add/afterpay-payment-method`)
- Run `npm install`, `npm run build:webpack` and start it with `npm run up`
- Connect your Stripe account (from any [supported countries](https://docs.stripe.com/payments/afterpay-clearpay)) with the Afterpay (Clearpay) payment method enabled in `/wp-admin/admin.php?page=wc-settings&tab=checkout&section=stripe&panel=methods`
- Enable the payment method on your Stripe dashboard
- Set your store currency to any of the [supported ones](https://docs.stripe.com/payments/afterpay-clearpay)
- Configure your webhook URL

### Display (merchant)

- As a merchant, go to the Stripe extension settings page `/wp-admin/admin.php?page=wc-settings&tab=checkout&section=stripe&panel=methods`
- Confirm that you can see the new payment method icon and details:
![Screenshot 2024-04-24 at 14 51 02](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/10187816/4b511495-68f0-4e97-88df-f0831d22fb53)
- Now, change your account country to the UK
- Confirm that the payment method was renamed to Clearpay with a new icon
![Screenshot 2024-04-24 at 15 22 40](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/10187816/c8a56a96-5a6d-48ce-b150-6b7c08340897)

### Display (shopper)

- As a shopper, add more than 4000 USD or any [other limited amount](https://docs.stripe.com/payments/afterpay-clearpay)
- Go to the checkout page
- Confirm that you cannot see the Afterpay payment method there
- Now, add less than that to your cart
- Go to the checkout page again
- Confirm that Afterpay is now showing up
![Screenshot 2024-04-22 at 09 50 08](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/10187816/7f31c4cf-d9fa-4347-b616-fa17b8c8a305)
- Repeat the same steps for the shortcode and block checkout
- And repeat using a Stripe account from the UK, confirming the new name and icon (Clearpay)

### Processing

- As a shopper, complete your checkout above
- In the Afterpay/Clearpay sandbox screen, create an account and use the test pin to confirm it (`123456`)
- Chose the installment options
- Confirm that you can see your order as `Processing` in wp-admin
- Repeat the tests with the pre-orders and subscriptions extensions

### Failures

- Repeat the tests above for a regular order
- In the Afterpay/Clearpay sandbox screen, click the "close button" at the top
- Confirm that your order is marked as `Failed` in wp-admin

### Refunds

- You may need to use ngrok to test this (or a store that is available on the internet)
- This should work only after https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3075 is merged (and the base branch updated)

#### Full

- Select any of the completed orders in the Stripe dashboard
- Refund it
- Go to your store and confirm that the order is now refunded

#### Partial

- Select any of the completed orders in the Stripe dashboard
- Refund it partially
- Go to your store and confirm that the order is now refunded (partially)

<!--
Please follow the following guidelines when writing testing instructions:

- Include screenshots if there is no similar flow in the critical flows: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Critical-flows
- Assume instructions will be copied over to the Release Testing Instructions wiki page: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions
- Assume instructions will be followed by external testers.
- Assume tester does not have intimate knowledge of Stripe.
-->

> ℹ️ Known issue: The Clearpay icon isn't displayed on the Block checkout. It'll be fixed in https://github.com/woocommerce/woocommerce-gateway-stripe/issues/3111

---

-   [x] Covered with tests (or have a good reason not to test in description ☝️)
-   [x] Added changelog entry **in both** `changelog.txt` and `readme.txt` (or does not apply)
-   [ ] Tested on mobile (or does not apply)

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)

a-danae committed:
Add domestic payment restrictions and multiple currency support for Klarna

a-danae committed:
Inject WC_Stripe_Account as a dependency of WC_Stripe_UPE_Payment_Gateway

wjrosa committed:
Starting the Affirm implementation

wjrosa committed:
Including common updates to shared classes

wjrosa committed:
Including temp payment icon

wjrosa committed:
Fix include error

wjrosa committed:
Updating currencies and icon

wjrosa committed:
Additional icon + copying some common unit test changes

wjrosa committed:
Additional icon + copying some common unit test changes

wjrosa committed:
Adding currency limit verification

wjrosa committed:
Fix conversion rate

wjrosa committed:
Fix currency conversion

wjrosa committed:
Fix payment method being displayed on the pay for order page

wjrosa committed:
Fix limit verification

wjrosa committed:
Adding specific unit tests

wjrosa committed:
Adding specific unit tests

wjrosa committed:
Fix tests

wjrosa committed:
Revert unnecessary change

wjrosa committed:
Readme and changelog updates

wjrosa committed:
Starting Afterpay method implementation

wjrosa committed:
Multiple updates to the main payment method classes

wjrosa committed:
Conditional attributes for Clearpay

wjrosa committed:
Introducing a new payment method data hook

wjrosa committed:
Fix include

wjrosa committed:
Improve conditional strings handling

wjrosa committed:
Adding specific unit tests

wjrosa committed:
Readme and changelog update + fix unit test

wjrosa committed:
Fix unit JS tests

wjrosa committed:
Fix tests

wjrosa committed:
Fix tests

wjrosa committed:
Fix tests

wjrosa committed:
Fix tests

wjrosa committed:
Add missing icon image

wjrosa committed:
Adding limits per currency

a-danae commented on client/utils/use-payment-methods-data.js:
Looks like this file is only intended to modify the Afterpay / Clearpay label, icon, and description depending on the account country. If so, would it simplify the code if we did this earlier in the payment-methods-map?
https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3091/files#diff-ea424360da29ad35e9247ee2d2550b8b2bcefc85e720b1a4592b6a0ee72a22b1R62-R77

This is how WooPayments is handling it.
https://github.com/Automattic/woocommerce-payments/blob/develop/client/payment-methods-map.tsx#L196-L220

Another advantage of handling this conditional early on is that we won't have to do this check in other places when displaying Afterpay / Clearpay information.

wjrosa replied:
I agree, but I tried that, and it did not work. For some reason, the map file is loaded before the account hook. Also, I could not call the file in this file (maybe because it is a JS, not a JSX file). But I will try again 👀

wjrosa committed:
Fix rebase issue

wjrosa committed:
Fix rebase issues

a-danae committed:
Switch the keys of the first and second levels of the currency limit array

a-danae committed:
Display the afterpay/clearpay icon on the block checkout

a-danae committed:
Define the display of afterpay/clearpay in the settings via JS params

a-danae committed:
Remove use-payment-methods-data

a-danae committed:
Access wc_stripe_settings_params through the window dom object

a-danae committed:
Adjust unnecesary diffs

a-danae committed:
Remove useAccount mocks

wjrosa commented on client/payment-methods-map.js:
```diff
@@ -1,6 +1,9 @@
 import icons from './payment-method-icons';
 
+const accountCountry =
+	window.wc_stripe_settings_params?.account_country || 'US';
```
Oh, this is much better!

wjrosa commented on includes/payment-methods/class-wc-stripe-upe-payment-method-afterpay-clearpay.php:
```diff
@@ -0,0 +1,125 @@
+	 *
+	 * @return boolean
+	 */
+	private function is_gb_country() {
```
I guess this is not needed anymore with your latest changes, @a-danae

a-danae replied:
Looks like we still change the title, description, and icon based on the country using this method. For example, the icon on the shortcode checkout changes based on this. Glad to learn if I'm missing something there, though!

wjrosa replied:
Oops, you're right. Sorry! I forgot about that part 😓

a-danae committed:
Update afterpay constant on the client side

a-danae committed:
Temporary fix for rendering Clearpay on the block checkout

Prevent the block checkout from breaking on render when there's a GB Stripe account connected. This displays the Afterpay icon instead of the Clearpay one. Fixing that separately to allow testing the rest of the functionality in the meantime.

a-danae committed:
Update innacurate inline comments on currency limits

a-danae committed:
Don't validate currency limit in non-checkout contexts

james-allan commented on includes/payment-methods/class-wc-stripe-upe-payment-method-afterpay-clearpay.php:
```diff
@@ -0,0 +1,125 @@
+	}
+
+	/**
+	 * /**
```
```suggestion

```

james-allan replied:
Fixed in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3091/commits/f2acb78aec96fa4c88083d0e79f7dc44514f4303

james-allan commented:
The Clearpay icon on the Block checkout uses the Afterpay icon. The difference is barely visible. 

<img width="1248" alt="Screenshot 2024-05-02 at 4 42 54 PM" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/41473e98-bcc5-4a2e-9817-8ea57eab9176">

I noticed it's flagged in the code as `// Temporary. There's WIP to fix this.` so I'm guessing you're aware of this.

Mayisha reviewed (approved):
Nice work here! With this PR we now have Afterpay/Clearpay available for the merchants :tada: 

All the test cases in happy path are successful and work as expected.

**Display**
:green_circle: When Stripe API keys are from a US based account and store currency is USD, Afterpay is available in the Stripe settings page to enable/disable. Afterpay is offered on the checkout page.
:green_circle: When Stripe API keys are from a UK based account and store currency is GBP, Clearpay is available in the Stripe settings page to enable/disable. Clearpay is offered on the checkout page.

**Payment processing**
:green_circle: Payment processing is successful with AfterPay from checkout page (US based Stripe account and store currency USD)
:green_circle: Payment processing is successful with AfterPay from pay for order page(US based Stripe account and store currency USD)
:green_circle: Payment processing is successful with ClearPay from checkout page (UK based Stripe account and store currency GBP)
:green_circle: Payment processing is successful with ClearPay from pay for order page (UK based Stripe account and store currency GBP)
:green_circle: If I simulate a failed transaction, the order status is set to `failed`.

**Refund**
:green_circle: Full refund works. The order in wp-admin is marked as refunded and there is a corresponding refund entry in the Stripe dashboard.
:green_circle: Partial refund works. The order in wp-admin is marked as processing with the updated price and there is a corresponding refund entry in Stripe dashboard.

I have got one question and a couple of enhancement suggestions.
:yellow_circle: When Stripe API keys are from a UK based account and store currency is USD, Clearpay is available in the Stripe settings page to enable/disable. But Clearpay is not present on the checkout page. Clearpay should be disabled in the settings page because of currency requirement with a tooltip saying the required currency is GBP.
:yellow_circle: When Stripe API keys are from a France based account and store currency is Euro, neither Afterpay is present in the Stripe settings page in disabled state (because of currency requirement). If I change the store currency to USD I can enable/disable Afterpay in the settings page but Afterpay is not present on the checkout page. Afterpay/Clearpay should not be present in the settings page if account country is not supported.
:question: On the order received page the `Payment Method` is `Clearpay / Afterpay` instead of just `Clearpay` or `Afterpay`. Is this expected?

![order received](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/33387139/53fdccb8-15ca-4996-a4e5-43738168ed39)

There is an icon related issue on the block checkout page that James discovered here https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3091#issuecomment-2089741696.

james-allan commented:
> When Stripe API keys are from a UK based account and store currency is USD, Clearpay is available in the Stripe settings page to enable/disable. But Clearpay is not present on the checkout page. Clearpay should be disabled in the settings page because of currency requirement with a tooltip saying the required currency is GBP.

This looks like similar feedback to what I received on the Affirm gateway PR here: https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3103#pullrequestreview-2034192358

I think we might be best to fix this across the board for all our domestic restricted Payment methods. Perhaps as part of a separate issue.

james-allan committed:
Remove /**

james-allan committed:
Use the correct icon for clearpay/afterpay based on Stripe Account country

james-allan commented:
@wjrosa I hope you don't mind but I submitted two changes to address the two issues I mentioned. Namely the icon issue on the checkout and https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3091#discussion_r1587123255. I'll also look into resolving the conflicts so we can move forward with this PR asap.

james-allan committed:
Merge branch 'develop' into add/afterpay-payment-method

james-allan committed:
Fix icon issues on Admin pages

james-allan commented:
> On the order received page the Payment Method is Clearpay / Afterpay instead of just Clearpay or Afterpay. Is this expected?

I think this is fine. Stripe themselves refer to Afterpay Clearpay in their dashboard:

<img width="509" alt="Screenshot 2024-05-03 at 4 15 37 PM" src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/8490476/8de36aab-eead-4a61-9219-53db06e41c63">

We have better control on the WooCommerce side to label this so I will file an issue just in case it is seen as a slight improvement to label it correctly.

james-allan commented:
> 🟡  When Stripe API keys are from a France based account and store currency is Euro, neither Afterpay is present in the Stripe settings page in disabled state (because of currency requirement). If I change the store currency to USD I can enable/disable Afterpay in the settings page but Afterpay is not present on the checkout page. Afterpay/Clearpay should not be present in the settings page if account country is not supported.
> ref: https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3091#pullrequestreview-2035789883

I've created this issue for that: https://github.com/woocommerce/woocommerce-gateway-stripe/issues/3112

> When Stripe API keys are from a UK based account and store currency is USD, Clearpay is available in the Stripe settings page to enable/disable. But Clearpay is not present on the checkout page. Clearpay should be disabled in the settings page because of currency requirement with a tooltip saying the required currency is GBP.
> ref: https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3091#pullrequestreview-2035789883

I believe this is the same issue you mentioned on the Affirm PR https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3103#issuecomment-2091961654. I don't know if we need to create an issue for that, but feel free to if you disagree that this is an issue.

james-allan committed:
Merge branch 'develop' into add/afterpay-payment-method
